### PR TITLE
Preparatory work before merging the ChainDB implementation

### PR DESCRIPTION
**Pull request by mrBliss**

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs [hunk 1]
```diff
@@ -224,8 +224,10 @@ toChain chainDB = bracket
     go chain it = do
       next <- iteratorNext it
       case next of
-        IteratorExhausted  -> return chain
-        IteratorResult blk -> go (Chain.addBlock blk chain) it
+        IteratorResult blk  -> go (Chain.addBlock blk chain) it
+        IteratorExhausted   -> return chain
+        IteratorBlockGCed _ ->
+          error "block on the current chain was garbage-collected"
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs]: In theory this can happen due to timing, even though in practice this is extrordinarily unlikely because this function is eager so will not trigger the GC corner case that requires one to use it very very slowly.

But since it's only used for tests, it's fine. Still, since people may copy'n'paste, no harm in using `throwM $ BlockGCed @blk hash` or whatever.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs [hunk 2]
```diff
@@ -272,6 +274,11 @@ newtype IteratorId = IteratorId Int
 data IteratorResult blk =
     IteratorExhausted
   | IteratorResult blk
+  | IteratorBlockGCed (HeaderHash blk)
+    -- ^ The block that was supposed to be streamed was garbage-collected from
+    -- the VolatileDB, but not added to the ImmutableDB.
+    --
+    -- This will only happen when streaming very old forks.
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs]: very old forks very slowly

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs [hunk 3]
```diff
@@ -286,6 +309,63 @@ data IteratorResult blk =
...
+-- To stream all blocks from the current chain from the ChainDB, one would use
+-- @'StreamFromExclusive' 'genesisPoint'@ as the lower bound and
+-- @'StreamToInclusive' tip@ as the upper bound where @tip@ is retrieved with
+-- 'getTipPoint'.
+--
+-- However, when the ChainDB is empty, @tip@ will be 'genesisPoint' too, in
+-- which case the bounds don't make sense. This function correctly handles
+-- this case.
+--
+-- Note that this is note a 'Reader', so the stream will not include blocks
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs]: note/not

Good clarification on the difference between a reader and an iterator: one very much live, the other as static as practically possible. And true, they could be confused. Perhaps we should put this note in the docs for the iterator and/or reader types.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs [hunk 4]
```diff
@@ -286,6 +309,63 @@ data IteratorResult blk =
...
+-- that are added to the current chain after starting the stream.
+streamAll :: (MonadSTM m, StandardHash blk) => ChainDB m blk
+          -> m (Maybe (Iterator m blk))
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs]: So we can't have an iterator that returns nothing? We need the `Maybe`?

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockFetchServer.hs [hunk 5]
```diff
@@ -57,30 +58,34 @@ blockFetchServer _tracer chainDB = senderSide
...
         IteratorResult blk     -> return $ SendMsgBlock blk (sendBlocks it)
         IteratorBlockGCed hash -> throwM $ BlockGCed @blk hash
 
-    withIter :: ChainRange blk -> (ChainDB.Iterator m blk -> m a) -> m a
+    withIter :: ChainRange blk
+             -> (Either (UnknownRange blk) (ChainDB.Iterator m blk) -> m a)
+             -> m a
     withIter (ChainRange start end) = bracket
       -- TODO when streamBlocks throws an exception, we should return
       -- SendMsgNoBlocks. Which exception will the ChainDB throw?
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/BlockFetchServer.hs]: This TODO looks outdated now. Now that we report range errors up front, when an iterator throws an exception later then we can just let that propagate.

File: ouroboros-consensus/src/Ouroboros/Storage/LedgerDB/InMemory.hs [hunk 6]
```diff
@@ -266,6 +271,55 @@ ledgerDbTail (Snap _ _ ss) = ledgerDbTail ss
...
+  -- | All blocks were valid, a 'LedgerDB' corresponding to the last valid
+  -- block is returned.
+  ValidBlocks ::            LedgerDB l r -> PushManyResult e l r ap
+
+-- | The result of 'ledgerDbSwitch', i.e. when validating a fork of the
+-- current chain.
+--
+-- First, we roll back @m@ blocks and then apply @n@ new blocks. It is
+-- important that @n >= m@, because we cannot return a valid 'LedgerDB' that
+-- is "older" (corresponds to an older block) than the given one.
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Storage/LedgerDB/InMemory.hs]: escape " in haddock markup.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs [hunk 7]
```diff
@@ -114,6 +115,18 @@ lastK (SecurityParam k) f =
...
+-- | The block number of the most recent \"immutable\" block, i.e. the oldest
+-- block we can roll back to. We cannot roll back the block itself.
+--
+-- In the real implementation this will correspond to the block number of the
+-- block at the tip of the Immutable DB.
+immutableBlockNo :: HasHeader blk
+                 => SecurityParam -> Model blk -> Block.BlockNo
+immutableBlockNo (SecurityParam k) =
+        Chain.headBlockNo
+      . Chain.drop (fromIntegral k)
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs]: I think that's right, no off-by-one errors as far as I can see, and clear doc comment to explain.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs [hunk 8]
```diff
@@ -374,3 +376,25 @@ between (SecurityParam k) from to m = do
         -> return frag'
         | otherwise
         -> Left $ MissingBlock p
+
+-- Would the garbage collector (if run) be able to garbage collect the block
+-- corresponding to the given point?
+garbageCollectable :: forall blk. HasHeader blk
+                   => SecurityParam -> Model blk -> Point blk -> Bool
+garbageCollectable (SecurityParam k) m@Model{..} p =
+    not onCurrentChain && olderThanK
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs]: :+1:

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs [hunk 9]
```diff
@@ -374,3 +376,25 @@ between (SecurityParam k) from to m = do
...
+  where
+    onCurrentChain = Chain.pointOnChain p (currentChain m)
+    olderThanK     = Chain.pointSlot    p < oldestAllowedSlot
+
+    oldestAllowedSlot = SlotNo $ slotAtTip - min k slotAtTip
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs]: This doesn't look right. This is taking k slots, not k blocks. Immutability / no-rollback is about K blocks, not slots.

What test would have caught this? e.g. in `garbageCollect` below we could have deleted forks that were only k/2 old. Perhaps we want a test that covers switching to a fork at the maximum possible point back, interleaved with GCing, and checking we can still switch since the blocks are still there. Just a thought. You can probably think of a better one, and note it down for later in one of the tickets.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs [hunk 10]
```diff
@@ -250,26 +251,48 @@ data IteratorResult blk =
     IteratorExhausted
   | IteratorResult blk
 
+{-------------------------------------------------------------------------------
+  Chain updates
+
+  This is a generalization of 'ChainUpdate' used in the network layer to
+  support a 'Functor' instance.
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs]: Good idea, will do.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs (hunk 9 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs]: Of course, silly me. Thanks for catching this. See #693.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs [hunk 11]
```diff
@@ -175,22 +176,24 @@ iteratorNext itrId m =
   Readers
 -------------------------------------------------------------------------------}
 
-newReader :: HasHeader blk => Model blk -> (CPS.ReaderId, Model blk)
-newReader m = (rdrId, m { cps = cps' })
+readBlocks :: HasHeader blk => Model blk -> (CPS.ReaderId, Model blk)
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Model.hs]: Right, that makes sense: #694.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs (hunk 4 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs]: Returning an empty iterator would be nicer, except for the fact that the returned iterator should have an `IteratorId` which is unique to the ChainDB (used for equality). Generating  such a unique `IteratorId` without *inside* knowledge of the ChainDB implementation is not easy.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs (hunk 1 above)
Comment (nc6) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/API.hs]: You have the `BlockGCed` error but don't seem to use it here?

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Mock.hs [hunk 12]
```diff
@@ -45,6 +44,9 @@ openDB cfg initLedger blockHeader = do
         query' :: (Model blk -> a) -> m a
         query' = atomically . query
 
+        queryE' :: (Model blk -> Either (ChainDbError blk) a) -> m a
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Mock.hs]: These names aren't especially helpful. Given the worries about not many people knowing the chainDB code, it might be nicer to have more explanatory names here.

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Mock.hs [hunk 13]
```diff
@@ -45,6 +44,9 @@ openDB cfg initLedger blockHeader = do
         query' :: (Model blk -> a) -> m a
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Mock.hs]: queryAtomically?

File: ouroboros-consensus/src/Ouroboros/Storage/LedgerDB/InMemory.hs [hunk 14]
```diff
@@ -268,6 +271,55 @@ ledgerDbTail (Snap _ _ ss) = ledgerDbTail ss
...
+{-------------------------------------------------------------------------------
+  Result types
+-------------------------------------------------------------------------------}
+
+-- | The result of 'ledgerDbPushMany', i.e. when validating a number of new
+-- blocks.
+--
+-- About the boolean type index: if all blocks given to 'ledgerDbPushMany' are
+-- 'Reapply', and so @ap ~ 'True@, then the result /must/ be 'ValidBlocks'.
+data PushManyResult e l r :: Bool -> * where
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Storage/LedgerDB/InMemory.hs]: Personally I find it easier to read as
```suggestion
data PushManyResult e l r (k :: Bool) where
```
But happy for this to be ignored!

File: ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/API.hs [hunk 15]
```diff
@@ -38,8 +38,8 @@ data VolatileDB blockId m = VolatileDB {
       -- | Return a function that returns the successors of the block with the
       -- given @blockId@ where 'Nothing' stands for Genesis.
       --
-      -- PRECONDITION: the block (if not Genesis) must be a member of the
-      -- VolatileDB, you can use 'getIsMember' to check this.
+      -- This function will return a non-empty set for any block of which a
+      -- predecessor has been added to the VolatileDB.
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/API.hs]: And will return an empty set if such a predecessor is not there...?

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Mock.hs (hunk 12 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Mock.hs]: The `BlockGCed` error is part of `BlockFetchServerException`, that's why I don't use it. But maybe I should rename it and use it here too.

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Mock.hs]: I'm not that worried about `queryE'` and so on being unhelpful. These are all local helpers with self-explanatory type signatures. If these names were longer, then `ChainDB` record implementation  would become less readable (too long lines), and this is the most important part.

File: ouroboros-consensus/src/Ouroboros/Storage/LedgerDB/InMemory.hs (hunk 14 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/LedgerDB/InMemory.hs]: I see. My thinking was: name type *parameters* but not type *indices*. Wasn't this a thing in Coq or Agda? I don't remember...
